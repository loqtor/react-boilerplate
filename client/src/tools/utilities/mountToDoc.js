/*
 * Utility to provide aXe with a fake DOM to allow for accessibility testing
 * on mounted components within our Jest tests
*/
import { mountWithIntl } from './enzymeIntlTestHelper';

let wrapper;

export default function mountToDoc(component) {
  const container = mountWithIntl(component);

  if (!wrapper) {
    wrapper = document.createElement('main');
    document.body.appendChild(wrapper);
  }

  wrapper.innerHTML = '';
  wrapper.appendChild(container.getDOMNode());

  return container;
}
