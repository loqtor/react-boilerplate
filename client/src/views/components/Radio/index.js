import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export class Radio extends Component {
  onChange = (event) => {
    const { onChange } = this.props;

    if (onChange) {
      onChange(event);
    }
  };

  render() {
    const {
      className,
      id,
      isChecked,
      isDisabled,
      isReadOnly,
      label,
      labelClassName,
      name,
      value,
    } = this.props;

    const inputClassName = classnames(className, { 'is-selected': isChecked });

    return (
      <Fragment>
        <label className={inputClassName} htmlFor={id}>
          <input
            className="Radio-input"
            disabled={isDisabled}
            id={id}
            onChange={this.onChange}
            readOnly={isReadOnly}
            type="radio"
            checked={isChecked}
            name={name}
            value={value}
          />
          <span className={labelClassName}>{label}</span>
        </label>
      </Fragment>
    );
  }
}

Radio.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  isChecked: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isReadOnly: PropTypes.bool,
  label: PropTypes.string.isRequired,
  labelClassName: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
    PropTypes.shape({}),
  ]),
};

Radio.defaultProps = {
  className: 'Radio',
  value: null,
  isChecked: false,
  isDisabled: false,
  onChange: null,
  labelClassName: 'Radio-label',
  isReadOnly: false,
};
