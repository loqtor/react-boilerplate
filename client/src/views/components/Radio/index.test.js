import React from 'react';
import { mount } from 'enzyme';
import axe from 'axe-core';
import toJson from 'enzyme-to-json';

import mountToDoc from '../../../tools/utilities/mountToDoc';

import Radio from './index';

const setup = (render, props) => {
  const onChangeSpy = jest.fn();

  const defaultProps = {
    id: 'Gender',
    label: 'Gender',
    name: 'Gender',
    value: 'male',
    onChange: onChangeSpy,
  };

  const component = render(<Radio {...defaultProps} {...props} />);

  return {
    actual: component,
    onChangeSpy,
  };
};

describe('<Radio />', () => {
  it('renders correctly', () => {
    const { actual } = setup(mount, {});

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when an error is present', () => {
    const { actual } = setup(mount, { error: 'Required field' });

    expect(toJson(actual)).toMatchSnapshot();
    expect(actual.find('.Radio-error').text()).toBe('Required field');
  });

  it('renders correctly when is disabled', () => {
    const { actual } = setup(mount, { isDisabled: true });

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('renders correctly when isReadOnly', () => {
    const { actual } = setup(mount, { isReadOnly: true });

    expect(toJson(actual)).toMatchSnapshot();
  });

  it('calls onChangeSpy correctly', () => {
    const { actual, onChangeSpy } = setup(mount, {});

    actual.find('input').simulate('change');
    expect(onChangeSpy).toHaveBeenCalledTimes(1);
  });

  it('has no accessibility violations', (done) => {
    const { actual } = setup(mountToDoc, {});
    const componentNode = actual.getDOMNode();

    axe.run(componentNode, (err, { violations }) => {
      expect(err).toBe(null);
      expect(violations).toHaveLength(0);
      done();
    });
  });
});
