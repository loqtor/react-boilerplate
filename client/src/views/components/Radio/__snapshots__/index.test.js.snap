// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> renders correctly 1`] = `
<Radio
  className="Radio"
  id="Gender"
  isChecked={false}
  isDisabled={false}
  isReadOnly={false}
  label="Gender"
  labelClassName="Radio-label"
  name="Gender"
  onChange={[MockFunction]}
  value="male"
>
  <label
    className="Radio"
    htmlFor="Gender"
  >
    <input
      checked={false}
      className="Radio-input"
      disabled={false}
      id="Gender"
      name="Gender"
      onChange={[Function]}
      readOnly={false}
      type="radio"
      value="male"
    />
    <span
      className="Radio-label"
    >
      Gender
    </span>
  </label>
</Radio>
`;

exports[`<Radio /> renders correctly when is disabled 1`] = `
<Radio
  className="Radio"
  id="Gender"
  isChecked={false}
  isDisabled={true}
  isReadOnly={false}
  label="Gender"
  labelClassName="Radio-label"
  name="Gender"
  onChange={[MockFunction]}
  value="male"
>
  <label
    className="Radio"
    htmlFor="Gender"
  >
    <input
      checked={false}
      className="Radio-input"
      disabled={true}
      id="Gender"
      name="Gender"
      onChange={[Function]}
      readOnly={false}
      type="radio"
      value="male"
    />
    <span
      className="Radio-label"
    >
      Gender
    </span>
  </label>
</Radio>
`;

exports[`<Radio /> renders correctly when isReadOnly 1`] = `
<Radio
  className="Radio"
  id="Gender"
  isChecked={false}
  isDisabled={false}
  isReadOnly={true}
  label="Gender"
  labelClassName="Radio-label"
  name="Gender"
  onChange={[MockFunction]}
  value="male"
>
  <label
    className="Radio"
    htmlFor="Gender"
  >
    <input
      checked={false}
      className="Radio-input"
      disabled={false}
      id="Gender"
      name="Gender"
      onChange={[Function]}
      readOnly={true}
      type="radio"
      value="male"
    />
    <span
      className="Radio-label"
    >
      Gender
    </span>
  </label>
</Radio>
`;
